README.md
# logentries-retrieving-log
[![NPM version][npm-image]][npm-url] [![Travis-CI Status][travis-image]][travis-url] [![Appveyor Status][appveyor-image]][appveyor-url] [![Daviddm Status][daviddm-image]][daviddm-url]

Retrieve logs from `logentries.com`. Unofficial.


## Install

```bash
$ npm install --save logentries-retrieving-log
```


## Usage

```javascript
var LogentriesRetrievingLog = require('logentries-retrieving-log');
var retriever = new LogentriesRetrievingLog({accountKey: '__YOUR_KEY__', logAddr: '__LOG_ADDR__'});
retriever.getLogs({limit: 10}, function(err, response, body){
  body.split("\n");
  //#=> [
  //  "145 <45>1 2014-12-16T13:00:45.518543+00:00 heroku api - - Add logentries:tryit add-on by o.gata.ken@gmail.com",
  //  "137 <45>1 2014-12-16T13:00:45.518543+00:00 heroku api - - Release v11 created by o.gata.ken@gmail.com",
  //  "129 <45>1 2014-12-16T13:00:46.043594+00:00 heroku web.1 - - State changed from up to starting",
  //  "131 <45>1 2014-12-16T13:00:48.084480+00:00 heroku web.1 - - Stopping all processes with SIGTERM",
  //  "124 <45>1 2014-12-16T13:00:49.345190+00:00 heroku web.1 - - Process exited with status 0",
  //  "155 <45>1 2014-12-16T13:00:51.140262+00:00 heroku web.1 - - Starting process with command `bin/hubot -a slack -n hubot`",
  //  "134 <190>1 2014-12-16T13:00:56.108557+00:00 app web.1 - - coffeelint@1.7.1 node_modules/coffeelint",
  //  "117 <190>1 2014-12-16T13:00:56.108575+00:00 app web.1 - - ├── ignore@2.2.15",
  //  "117 <190>1 2014-12-16T13:00:56.108577+00:00 app web.1 - - ├── resolve@0.6.3",
  //  "152 <190>1 2014-12-16T13:00:56.108578+00:00 app web.1 - - ├── optimist@0.6.1 (wordwrap@0.0.2, minimist@0.0.10)",
  //  ""
  // ]
});
```

## API

### LogentriesRetrievingLog.getLogs([options])

You can set `start`, `end`, `filter` and `limit`. See below logentries' document.

## Documentation for Logentries

* [API Downloads Documentation](https://logentries.com/doc/api-download/)
* [Account Key API Documentation](https://logentries.com/doc/accountkey/)


## Contributing

Welcome.


## License

Copyright (c) 2014 sanemat. Licensed under the MIT license.


[travis-url]: https://travis-ci.org/sanemat/node-logentries-retrieving-log
[travis-image]: https://img.shields.io/travis/sanemat/node-logentries-retrieving-log/master.svg?style=flat-square&label=travis
[appveyor-url]: https://ci.appveyor.com/project/sanemat/node-logentries-retrieving-log/branch/master
[appveyor-image]: https://img.shields.io/appveyor/ci/sanemat/node-logentries-retrieving-log/master.svg?style=flat-square&label=appveyor
[npm-url]: https://npmjs.org/package/logentries-retrieving-log
[npm-image]: https://img.shields.io/npm/v/logentries-retrieving-log.svg?style=flat-square
[daviddm-url]: https://david-dm.org/sanemat/node-logentries-retrieving-log
[daviddm-image]: https://img.shields.io/david/sanemat/node-logentries-retrieving-log.svg?style=flat-square


CHANGELOG.md
<a name="1.0.0"></a>
# [1.0.0](https://github.com/sanemat/node-logentries-retrieving-log/compare/v0.1.1...v1.0.0) (2015-09-10)

* Update dependencies


# 0.1.1 (2014-12-18)

- Treat getLogs params as option
- Remove Initial slash and trail slash

# 0.1.0 (2014-12-18)

- Retrieving log


test/logentries-retrieving-log_test.js
'use strict';

var LogentriesRetrievingLog = require('../');
var assert = require('power-assert');
var nock = require('nock');

describe('logentriesRetrievingLog', function () {
  var retriever;

  before(function(done){
    retriever = LogentriesRetrievingLog({ accountKey: '__YOUR_KEY__', logAddr: '__LOG_ADDR__' });
    done();
  });

  it('should retrieve logs length', function (done) {
    // https://pull.logentries.com/__YOUR_KEY__/__LOG_ADDR__/?limit=10
    nock('https://pull.logentries.com')
      .get('/__YOUR_KEY__/__LOG_ADDR__/?limit=10')
      .replyWithFile(200, __dirname + '/logentries-com-retrieved-logs.txt');
    retriever.getLogs({limit: 10}, function(err, response, body){
      assert.equal(body.trim().split("\n").length, 10);
      done();
    });
  });

  it('should retrieve logs contents', function (done) {
    // https://pull.logentries.com/__YOUR_KEY__/__LOG_ADDR__/?limit=10
    nock('https://pull.logentries.com')
      .get('/__YOUR_KEY__/__LOG_ADDR__/?limit=10')
      .replyWithFile(200, __dirname + '/logentries-com-retrieved-logs.txt');
    retriever.getLogs({limit: 10}, function(err, response, body){
      assert.equal(
        body.trim().split("\n")[3],
        "131 <45>1 2014-12-16T13:00:48.084480+00:00 heroku web.1 - - Stopping all processes with SIGTERM"
      );
      done();
    });
  });

  it('should retrieve logs contents without option', function (done) {
    // https://pull.logentries.com/__YOUR_KEY__/__LOG_ADDR__/
    nock('https://pull.logentries.com')
      .get('/__YOUR_KEY__/__LOG_ADDR__/')
      .replyWithFile(200, __dirname + '/logentries-com-retrieved-logs.txt');
    retriever.getLogs(function(err, response, body){
      assert.equal(
        body.trim().split("\n")[3],
        "131 <45>1 2014-12-16T13:00:48.084480+00:00 heroku web.1 - - Stopping all processes with SIGTERM"
      );
      done();
    });
  });

  it('should handle error response', function (done) {
    // https://pull.logentries.com/__YOUR_KEY__/__LOG_ADDR__/?invalid=true
    nock('https://pull.logentries.com')
      .get('/__YOUR_KEY__/__LOG_ADDR__/?invalid=true')
      .replyWithFile(404, __dirname + '/logentries-com-error-response.txt');
    retriever.getLogs({invalid: true}, function(err, response, body){
      assert.equal(!err, false);
      assert.equal(JSON.parse(body).response, 'error');
      done();
    });
  });
});


test/node-logentries-retrieving-log_test.js
'use strict';

var nodeLogentriesRetrievingLog = require('../');
var assert = require('power-assert');

describe('nodeLogentriesRetrievingLog', function () {

  it('should be awesome', function () {
    assert(nodeLogentriesRetrievingLog() == 'awesome');
  });

});


gulpfile.js
'use strict';

var gulp   = require('gulp');
var plugins = require('gulp-load-plugins')();
require('intelli-espower-loader');

var paths = {
  lint: ['./gulpfile.js', './lib/**/*.js'],
  watch: ['./gulpfile.js', './lib/**', './test/**/*.js', '!test/{temp,temp/**}'],
  tests: ['./test/**/*.js', '!test/{temp,temp/**}'],
  source: ['./lib/*.js']
};

var plumberConf = {};

if (process.env.CI) {
  plumberConf.errorHandler = function(err) {
    throw err;
  };
}

gulp.task('lint', function () {
  return gulp.src(paths.lint)
    .pipe(plugins.jshint('.jshintrc'))
    .pipe(plugins.plumber(plumberConf))
    .pipe(plugins.jscs())
    .pipe(plugins.jshint.reporter('jshint-stylish'));
});

gulp.task('istanbul', function (cb) {
  gulp.src(paths.source)
    .pipe(plugins.istanbul()) // Covering files
    .on('finish', function () {
      gulp.src(paths.tests)
        .pipe(plugins.plumber(plumberConf))
        .pipe(plugins.mocha())
        .pipe(plugins.istanbul.writeReports()) // Creating the reports after tests runned
        .on('finish', function() {
          process.chdir(__dirname);
          cb();
        });
    });
});

gulp.task('watch', ['test'], function () {
  gulp.watch(paths.watch, ['test']);
});

gulp.task('test', ['lint', 'istanbul']);

gulp.task('default', ['test']);


lib/logentries-retrieving-log.js
/*
 * logentries-retrieving-log
 * https://github.com/sanemat/node-logentries-retrieving-log
 *
 * Copyright (c) 2014 sanemat
 * Licensed under the MIT license.
 */

var request = require('request');
var fmt = require('util').format;
var trimmer = require('trimmer');

function LogentriesRetriever(credentials) {
  'use strict';

  if (!(this instanceof LogentriesRetriever)) {
    return new LogentriesRetriever(credentials);
  }
  credentials = credentials || {};
  this.accountKey = this._normalizeCredential(credentials.accountKey) || '';
  this.logAddr = this._normalizeCredential(credentials.logAddr) || '';
  this.apiHost = this._normalizeCredential(credentials.apiHost) || 'https://pull.logentries.com';
}

LogentriesRetriever.prototype._normalizeCredential = function(key) {
  if (typeof key === 'string' || key instanceof String) {
    return trimmer(key, '/');
  } else {
    return key;
  }
};

LogentriesRetriever.prototype.getLogs = function(params, callback) {
  if (typeof params === 'function'){
    callback = params;
    params = {};
  }
  params = params || {};

  var requestUrl = fmt('%s/%s/%s/', this.apiHost, this.accountKey, this.logAddr);
  request({url: requestUrl, qs: params}, function(err, response, body){
    if (err || response.statusCode !== 200){
      err = err || 'response.statusCode not eq 200';
      callback(err, response, body);
      return;
    }
    callback(null, response, body);
  });
};

module.exports = LogentriesRetriever;


example/simple.js
/*
 * logentries-retrieving-log
 * https://github.com/sanemat/node-logentries-retrieving-log
 *
 * Copyright (c) 2014 sanemat
 * Licensed under the MIT license.
 */

'use strict';

var LogentriesRetrievingLog = require('../');
var retriever = new LogentriesRetrievingLog({accountKey: '__YOUR_KEY__', logAddr: '__LOG_ADDR__'});
retriever.getLogs({limit: 10}, function(err, response, body){
  body.split("\n");
  //#=> [
  //  "145 <45>1 2014-12-16T13:00:45.518543+00:00 heroku api - - Add logentries:tryit add-on by o.gata.ken@gmail.com",
  //  "137 <45>1 2014-12-16T13:00:45.518543+00:00 heroku api - - Release v11 created by o.gata.ken@gmail.com",
  //  "129 <45>1 2014-12-16T13:00:46.043594+00:00 heroku web.1 - - State changed from up to starting",
  //  "131 <45>1 2014-12-16T13:00:48.084480+00:00 heroku web.1 - - Stopping all processes with SIGTERM",
  //  "124 <45>1 2014-12-16T13:00:49.345190+00:00 heroku web.1 - - Process exited with status 0",
  //  "155 <45>1 2014-12-16T13:00:51.140262+00:00 heroku web.1 - - Starting process with command `bin/hubot -a slack -n hubot`",
  //  "134 <190>1 2014-12-16T13:00:56.108557+00:00 app web.1 - - coffeelint@1.7.1 node_modules/coffeelint",
  //  "117 <190>1 2014-12-16T13:00:56.108575+00:00 app web.1 - - ├── ignore@2.2.15",
  //  "117 <190>1 2014-12-16T13:00:56.108577+00:00 app web.1 - - ├── resolve@0.6.3",
  //  "152 <190>1 2014-12-16T13:00:56.108578+00:00 app web.1 - - ├── optimist@0.6.1 (wordwrap@0.0.2, minimist@0.0.10)",
  //  ""
  //]
});

// API Downloads Documentation| Logentries Log Management Tools
// https://logentries.com/doc/api-download/
// Account Key API Documentation| Logentries Log Management
// https://logentries.com/doc/accountkey/
